static/progressive-web-app.js

```diff
@@ -12,77 +12,89 @@ const urlsToCache = [
+      } catch (error) {
+        console.error("[Service Worker] Error during activation:", error);
+      }
+      self.clients.claim(); // Take control of all pages immediately
+    })()
   );
-  self.clients.claim(); // Take control of all pages immediately
 });
 
-// Fetch event: Cache first approach but update cache anyways
+// Fetch event: Cache first approach but update cache anyway
 self.addEventListener("fetch", (event) => {
   const url = new URL(event.request.url);
 
   // Ignore non-HTTP(S) requests (like 'chrome-extension://')
-  if (url.protocol !== "http:" && url.protocol !== "https:") {
-    return;
-  }
+  if (url.protocol !== "http:" && url.protocol !== "https:") return;
 
   // If the request has query parameters, bypass the cache
-  if (url.search) { 
+  if (url.search) {
     event.respondWith(fetch(event.request));
     return;
   }
 
   event.respondWith(
-    caches.match(event.request).then((cachedResponse) => {
-      const fetchPromise = fetch(event.request)
-        .then((networkResponse) => {
-          // Clone the network response to avoid using the body twice
-          const responseClone = networkResponse.clone();
+    (async () => {
+      try {
+        const cachedResponse = await caches.match(event.request);
 
-          // If the network response is valid, update the cache
-          if (networkResponse.ok) {
-            caches.open(cacheName).then((cache) =>
-              cache.put(event.request, responseClone)
+        const fetchPromise = fetch(event.request)
+          .then(async (networkResponse) => {
+            if (networkResponse.ok) {
+              const responseClone = networkResponse.clone();
+              const cache = await caches.open(cacheName);
+              await cache.put(event.request, responseClone);
+            }
+            return networkResponse;
+          })
+          .catch((error) => {
+            console.error("[Service Worker] Network request failed:", error);
+            return (
+              cachedResponse ||
+              new Response("Offline content unavailable", {
+                status: 503,
+                statusText: "Service Unavailable",
+              })
             );
-          }
-          return networkResponse;
-        })
-        .catch((error) => {
-          console.error("[Service Worker] Network request failed:", error);
-          return cachedResponse || new Response("Offline content unavailable", {
-            status: 503,
-            statusText: "Service Unavailable",
           });
-        });
 
-      // Serve from cache first, but update the cache in the background
-      return cachedResponse || fetchPromise;
-    })
+        return cachedResponse || (await fetchPromise);
+      } catch (error) {
+        console.error("[Service Worker] Error handling fetch:", error);
+        return new Response("An error occurred", {
+          status: 500,
+          statusText: "Internal Server Error",
+        });
+      }
+    })()
```

Comment:
It's strange that you do self invoking functions to do async. Just wrap the entire callback inside an async function like we discussed here and then you can await them like normal. 

Also change catch to try catch. And don't use thenable, use async await.

Reply:
The self invoking functions are needed because event's `waitUntil()` method receives a promise as a parameter - not a function that returns a promise. ([See reference](https://www.w3.org/TR/service-workers/#wait-until-method)) I can only imagine that your idea was to remore the use of `waitUntil` entirely, for instance:

```
// Install event (caches all necessary files)
self.addEventListener("install", async (event) => {
  try {
    const cache = await caches.open(cacheName);
    await cache.addAll(urlsToCache);
    console.log("[Service Worker] Files cached successfully");
  } catch (error) {
    console.error("[Service Worker] Cache failed:", error);
  }
  self.skipWaiting(); // Activate the new worker immediately
});
```

instead of:

```
// Install event (caches all necessary files)
self.addEventListener("install", async (event) => {
  event.waitUntil(
    (async () => {
      try {
        const cache = await caches.open(cacheName);
        await cache.addAll(urlsToCache);
        console.log("[Service Worker] Files cached successfully");
      } catch (error) {
        console.error("[Service Worker] Cache failed:", error);
      }
      self.skipWaiting(); // Activate the new worker immediately
    })()
  );
});
```

The difference between these approaches is that with `waitUntil()` adding assets to cache becomes an asynchronous task, which would allow the task to be completed even after returning inside the event handler. During install this is not a problem, however in a fetch event I want the service worker to return whatever is in cache, but fire an asynchronous task to update the cache with a network response. If we remove `waitUntil()` and just await the network response, we would have to wait for network response to return whatever is in cache, seriously compromising speed in very slow networks. 

Here is a QA using `waitUntil()` to return from cache and asynchronously update cache with network response, in a simulated 10kbps network speed:

[10kbps-network-waitUntil.webm](https://github.com/user-attachments/assets/7429f941-b396-4134-bd0c-4b77726ac743)

Here is a QA awaiting the network response, in a simulated 10kbps network speed:

[10kbps-network-await.webm](https://github.com/user-attachments/assets/344132a9-8d2a-4054-a1dd-58943063792a)

Here are some additional references to clarify:

[What does event.waitUntil do in service worker and why is it needed?](https://stackoverflow.com/questions/37902441/what-does-event-waituntil-do-in-service-worker-and-why-is-it-needed)

[JavaScript how to use event.waitUntil() to wait for an async function call](https://stackoverflow.com/questions/65530092/javascript-how-to-use-event-waituntil-to-wait-for-an-async-function-call)

[Should I use event.waitUntil() or await fetch() or just fetch()?](https://community.cloudflare.com/t/should-i-use-event-waituntil-or-await-fetch-or-just-fetch/56183)

Reply:
Okay seems like you did your research so let's do what you think is right.

Reply:
That was good feedback, it's very important that we understand each others code practices, as well as document weird patterns. Thanks.